File: satfilt.m
t%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% This function filters the data of the satellites presented in the output.db
% file expecting it to be a CSV file. The criterion to do it is the following:
% path - Path where it will find the output.db file and where it will store the
%   resulting .db file.
% n_param - Parameter that is expected to filter:
%   2  - inclination (degrees)
%   3  - eccentricity
%   4  - period (minutes)
%   5  - perigee (degrees)
%   6  - longitude of the ascending node (degrees)
%   7  - mean anomaly (degrees)
%   -1 - all of them 
% max_val and min_val are the maximum and minimum values for the parameter 
%   selected. (If all the parameteres are being filtered or they are out of 
%   range, they will be computed automatically)
% num_groups - Number of satellites that you desire to achieve after the 
%   filtering (it may be less). If its value is different from -1, the group 
%   list will be ignored. 
% groups_lims - Array with the limits a the group as an array.
%   E.g.: [[15,30];[45,65];...;[329, 354]] 
%   The program will not controle if they are disjount groups or not.
%   If there is a number of groups specified, this option will be ignored.
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

function [] = satfilt(path, n_param, min_val, max_val, num_groups, groups_lims)

    search_path = strcat(path, "output.db");
    csvfile = dir(search_path);
    
    if numel(csvfile) >= 1
   
        printf("Loading data from %s\n", strcat(path, csvfile.name));
        data = csvread(strcat(path, csvfile.name))(2:end,:);
        
        % dynamic range
        dr(1) = 0;
        for k = 2:7
            dr(k) = max(data(:,k)) - min(data(:,k));
        end
        
        if n_param == -1
            n_min = 2;
            n_max = 7;
            min_val = 0;
            max_val = 1e6; % Infinitum
        else 
            n_min = n_param;
            n_max = n_param;
        end
        
        for n_param = n_min:n_max
        
            % compute max and min values
            max_val = min(max(data(:,n_param)), max_val);
            min_val = max(min(data(:,n_param)), min_val);
            
            % find a representative group for the parameter
            if num_groups == -1
                for i = 1:length(groups_lims)
                    num_groups = size(groups_lims,2);
                    groups_sat{i} = data(find(data(:,n_param) >= groups_lims(i,1) ...
                                           & data(:,n_param) <= groups_lims(i,2)),:);
                    groups_err{i} = zeros(1, size(groups_sat{i},1));
                end
            else
                delta_val = (max_val - min_val) / num_groups;
                for i = 1:num_groups
                    groups_sat{i} = data(find(data(:,n_param) >= (min_val + delta_val * (i-1)) ...
                                           & data(:,n_param) <= (min_val + delta_val * i)),:);
                    groups_err{i} = zeros(1, size(groups_sat{i},1));
                end 
            end

            % find the error of each value in each group
            for i = 1:num_groups
                for j = 1:num_groups
                    if i ~= j
                        for ii = 1:(size(groups_sat{i},1)-1)
                            for jj = 1:(size(groups_sat{j},1)-1)
                                aux = 0;
                                for k = 2:7
                                    if k ~= n_param
                                    aux = aux + abs(groups_sat{i}(ii,k) - ...
                                                   groups_sat{j}(jj,k)) / dr(k);
                                    end
                                end
                                groups_err{j}(jj) = groups_err{j}(jj) + aux;
                            end
                        end
                    end
                end  
            end
            
            % create a destination file
            if n_param == 2
                name = "inclination";
            elseif n_param == 3
                name = "eccentricity";
            elseif n_param == 4
                name = "period";
            elseif n_param == 5
                name = "perigee";
            elseif n_param == 6
                name = "longitude-ascending-node";
            else
                name = "mean-anomaly";
            end 
            out_path = strcat(path, name, ".db");
            out_file=fopen(out_path,'w');
            
            % write the header of the file
            fprintf(out_file, "NORAD's ID # inc ecc period perig long_a_n m_an \n");
                
            % find the best value of each group and write it in the file
            for i = 1:num_groups
                opt = groups_sat{i}(find(min(groups_err{i}) == groups_err{i}),:);
                if ~(isempty(opt))
                    fprintf(out_file, "%10.0f # ", opt(1));
                    for j = 2:7
                        fprintf(out_file, "%10.6f ", opt(j));
                    end
                    fprintf(out_file, "\n");
                end
            end  
            
            fclose(out_file);
            
            % reset values
            min_val = 0;
            max_val = 1e6; % Infinitum
            
        end
    
    else
        disp("No output.db file found")
    end

end

File: satfilt.cpp
/*************************************************************************************************
 *  \brief      Satellite filter through determinant parameters.
 *  \details    A tool to generate a database of satellites with relevant parameters.
 *  \author     Borja Rodríguez (borjabrg12@gmail.com)
 *  \version    0.1
 *  \date       27-oct-2017
 *  \copyright  GNU Public License (v3). This files are part of an on-going non-commercial research
 *              project at NanoSat Lab (http://nanosatlab.upc.edu) of the Technical University of
 *              Catalonia - UPC BarcelonaTech. Third-party libraries used in this framework might be
 *              subject to different copyright conditions.
 **************************************************************************************************/

 /*** INCLUDE SECTION ******************************************************************************/
#include "satfilt.hpp"

using namespace std;

/*** CONSTANTS ************************************************************************************/

#define DBG_REDB        "\x1b[31;1m"
#define DBG_REDD        "\x1b[31m"
#define DBG_GREENB      "\x1b[32;1m"
#define DBG_GREEND      "\x1b[32m"
#define DBG_BLUE        "\x1b[34;1m"
#define DBG_YELLOWB     "\x1b[33;1m"
#define DBG_YELLOWD     "\x1b[33m"
#define DBG_WHITEB      "\x1b[37;1m"
#define DBG_GREY        "\x1b[30;1m"
#define DBG_NOCOLOR     "\x1b[0m"

void printHelp(void)
{
    /*  OPTION      VALUE       DESCRIPTION:
     *  -t          folder path Path to the TLE folder.
     *  -o          folder paht Path to the results folder.
     *  -v          (none)      Verbose: outputs all data points as it generates them.
     *  -h          (none)      Shows the help menu.
     */
    cout << "List of possible arguments:" << endl;
    cout << DBG_WHITEB "OPTION   VALUE                 DESCRIPTION" DBG_NOCOLOR << endl;
    cout << DBG_REDD   "  -t     " DBG_YELLOWD "Path to TLE folder    " DBG_NOCOLOR "Path to a folder containing (only) Two-Line Elements collection files." << endl;
    cout << DBG_REDD   "  -o     " DBG_YELLOWD "Path to folder        " DBG_NOCOLOR "Path to the results folder (if it doesn't exist, it'll be created)." << endl;
    cout << DBG_REDD   "  -v     " DBG_YELLOWD "(none)                " DBG_NOCOLOR "Verbose; will output all data points as it generates them." << endl;
    cout << DBG_REDD   "  -h     " DBG_YELLOWD "(none)                " DBG_NOCOLOR "Shows this help." << endl;
}

void printHeader(bool first)
{
    if(first) {
        cout << "┏━━━━━━━━┳━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━┓" << endl;
    } else {
        cout << "┢━━━━━━━╈━━━━━━━━╈━━━━━━━━━━━╈━━━━━━━━━━━╈━━━━━━━━━━━╈━━━━━━━━━━━━╈━━━━━━━━━━━┪" << endl;
    }
    cout << "┃   ID   ┃  inc   ┃    ecc    ┃     T     ┃    per    ┃  long_a_n  ┃   m_an    ┃" << endl;
    cout << "╇━━━━━━━━╇━━━━━━━━┃━━━━━━━━━━━┃━━━━━━━━━━━┃━━━━━━━━━━━┃━━━━━━━━━━━━┃━━━━━━━━━━━┃" << endl;
}

void printFooter(void)
{
    cout << "╇━━━━━━━━╇━━━━━━━━┃━━━━━━━━━━━┃━━━━━━━━━━━┃━━━━━━━━━━━┃━━━━━━━━━━━━┃━━━━━━━━━━━┃" << endl;
}

int main(int argc, char **argv){

    /* C-like variable declaration is chosen here for documentation and code cleanliness purposes.*/
    string input_path;          /* Input path where TLE files are located.                        */
    string tle_file_path;       /* The TLE file name/path.                                        */
    string output_path_root;    /* Path folder for the resulting files.                           */
    string output_path;         /* Output path for the resulting files.                           */
    DIR * tle_directory;        /* TLE directory.                                                 */
    struct dirent * dir_entry;  /* File within the TLE directory.                                 */
    FILE * tle_file;            /* TLE input file.                                                */
    FILE * output_file;         /* Resulting output file.                                         */
    char file_line[200];        /* One single line from an open file.                             */
    int line_count = 0;         /* Line counter (debug purposes).                                 */
    string tle_line1;           /* Line 1 in the TLE file.                                        */
    string tle_line2;           /* Line 2 in the TLE file.                                        */
    string tle_line3;           /* Line 3 in the TLE file.                                        */
    time_t time_start;          /* In seconds                                                     */
    char time_formated[21];     /* Time in the format "yyyy-mm-dd hh:mm:ss"                       */
    struct tm *tmp;             /* Time struct (debug purposes).                                  */
    bool verbose = false;       /* Whether to print all data points as they are generated.        */


    /* Internal set-up: ------------------------------------------------------------------------- */
    char sat_name[25];          /* Name of the satellite                                          */
    double norad;               /* Line 3 in the TLE file.                                        */
    double inc;                 /* Inclination degrees.                                           */
    double ecc;                 /* Eccentricity of the /(no units)                                */
    double period;              /* Period of one orbit (minutes)                                  */
    double perig;               /* Argument of the perigee (degrees)                              */
    double long_a_n;            /* Longitude of the ascending node (degrees)                      */
    double m_an;                /* Mean anomaly (degrees)                                         */

    /* Setup the default configuration values: */
    time_start = time(NULL);
    tmp = localtime(&time_start);
    strftime(time_formated, 21, "%Y-%m-%d_%H%M%S", tmp);
    output_path_root = "database/" + string(time_formated);
    input_path = "tle_collections";

    /* Configuration based on program arguments: ------------------------------------------------ */
    if(argc >= 1)
    {
        /*  OPTION      VALUE       DESCRIPTION:
         *  -t          folder path Path to the TLE folder.
         *  -o          folder paht Path to the results folder.
         *  -v          (none)      Verbose: outputs all data points as it generates them.
         *  -h          (none)      Shows the help menu.
         */
        int arg_iterator = 1;
        while(arg_iterator < argc)
        {
            string str = string(argv[arg_iterator]);
            if(str == "-t" && (arg_iterator + 1) < argc)
            {
                input_path = string(argv[arg_iterator + 1]);
                arg_iterator++;
            }
            else if(str == "-o" && (arg_iterator + 1) < argc)
            {
                output_path_root = string(argv[arg_iterator + 1]) + "/" + string(time_formated);
                arg_iterator++;
            }
            else if(str == "-v")
            {
                verbose = true;
            }
            else
            {
                cout << "Unknown argument: \'" << str << "\'" << endl;
                printHelp();
                return -1;
            }
            arg_iterator++;
        }

        /* Create results folder: */
        system(string("mkdir -p " + output_path_root).c_str());     /* Linux/Bash-specific.       */

        /* Creating output file */
        output_path = output_path_root + "/output" + ".db";
        if((output_file = fopen(output_path.c_str(), "w+")) == NULL) {
            cerr << DBG_REDD "Unable to open file " << output_path << DBG_NOCOLOR << endl;
        }

        /* Write CSV headers to the output file. */
        fprintf(output_file, "NORAD's ID, inc, ecc, period, perig, long_a_n, m_an \n");

        if(verbose){
            printHeader(true);
        }

        /* Entering in the tle_directory. */
        if((tle_directory = opendir(input_path.c_str())) != NULL) {

            while((dir_entry = readdir(tle_directory)) != NULL) {
                if(strcmp(dir_entry->d_name, ".") && strcmp(dir_entry->d_name, "..")) {
                    string full_path = input_path + "/" + string(dir_entry->d_name);
                    if((tle_file = fopen(full_path.c_str(), "r")) != NULL) {
                       while(fgets(file_line, 80, tle_file) != NULL) {
                            if(sscanf(file_line, "%24[^\n\t\r]", sat_name) < 1) {
                                cerr << "Can't read satellite name, will try to continue."  << endl;
                            } else {
                                tle_line1 = string(file_line);
                            }
                            line_count++;
                            if(fgets(file_line, 80, tle_file) != NULL) {
                                tle_line2 = string(file_line);
                            } else {
                                break;
                            }
                            line_count++;
                            if(fgets(file_line, 80, tle_file) != NULL) {
                                tle_line3 = string(file_line);
                            } else {
                                break;
                            }
                            line_count++;

                            /* Getting satellite information. */
                            norad = stod(tle_line2.substr(3,5));
                            inc = stod(tle_line3.substr(9,8));
                            ecc = stod(tle_line3.substr(26,7)) / 10000000;
                            period = (1 / stod(tle_line3.substr(52,11))) * 24 * 60;
                            perig = stod(tle_line3.substr(34,8));
                            long_a_n = stod(tle_line3.substr(17,8));
                            m_an = stod(tle_line3.substr(43,8));

                            /*Writting CSV file. */
                            fprintf(output_file, "%10.0f,", norad);
                            fprintf(output_file, "%.3f,%.6f,%.3f,%.3f,%.3f,%.3f \n",
                            inc, ecc, period, perig, long_a_n, m_an);

                            if(verbose){
                                printf("┃ %6.0f ┃ %6.3f ┃ %9.6f ┃ %9.3f ┃ %9.3f ┃  %9.3f ┃ %9.3f ┃ \n",
                                norad, inc, ecc, period, perig, long_a_n, m_an);
                            }
                        }
                    }
                }
            }

            if(verbose){
                printFooter();
            }

        }

    }
}
